src/server/pegasus_hotkey_collector.cpp

```diff
@@ -0,0 +1,279 @@
+                                    "coarse_capture_hash_bucket_num",
+                                    37,
+                                    "the number of coarse capture hash buckets");
+
+    data_variance_threshold =
+        dsn_config_get_value_uint64("pegasus.server",
+                                    "data_variance_threshold",
+                                    3,
+                                    "the threshold of variance_cal to find the outliers");
+
+    _collector_start_time = dsn_now_s();
+
+    hotbucket_hash = [](const std::string &data, const int split_count) {
+        return (int)(std::hash<std::string>{}(data) % split_count);
+    };
+}
+
+bool hotkey_collector::init()
+{
+    switch (_collector_state.load()) {
+    case COARSE:
+        ddebug("Now is detecting %s hotkey, state is %s",
+               hotkey_type == dsn::apps::hotkey_type::READ ? "read" : "write",
+               "COARSE");
+        return false;
+    case FINE:
+        ddebug("Now is detecting %s hotkey, state is %s",
+               hotkey_type == dsn::apps::hotkey_type::READ ? "read" : "write",
+               "FINE");
+        return false;
+    case FINISH:
+        ddebug(
+            "%s hotkey result has been found, you can send a stop rpc to restart hotkey detection",
+            hotkey_type == dsn::apps::hotkey_type::READ ? "Read" : "Write");
+        return false;
+    case STOP:
+        _collector_start_time = dsn_now_s();
+        _coarse_data_collector.reset(new hotkey_coarse_data_collector(this));
+        _collector_state.store(COARSE);
+        ddebug("Is starting to detect %s hotkey",
+               hotkey_type == dsn::apps::hotkey_type::READ ? "read" : "write");
+        return true;
+    default:
+        derror_replica("wrong collector state");
+        return false;
+    }
+}
+
+void hotkey_collector::clear()
+{
+    _collector_state.store(STOP);
+    _coarse_data_collector.reset();
+    _fine_data_collector.reset();
+    derror_replica("Already cleared {} hotkey cache",
+                   hotkey_type == dsn::apps::hotkey_type::READ ? "read" : "write");
+}
+
+std::string hotkey_collector::get_status()
+{
+    switch (_collector_state.load()) {
+    case COARSE:
+        return "COARSE";
+    case FINE:
+        return "FINE";
+    case FINISH:
+        return "FINISH";
+    case STOP:
+        return "STOP";
+    default:
+        derror_replica("wrong collector state");
+        return "false";
+    }
+}
+
+bool hotkey_collector::get_result(std::string &result)
+{
+    if (_collector_state.load() != FINISH)
+        return false;
+    result = _fine_result;
+    return true;
+}
+
+void hotkey_collector::capture_msg_data(dsn::message_ex **requests, const int count)
+{
+    if (_collector_state.load() == STOP || _collector_state.load() == FINISH || count == 0) {
+        return;
+    }
+    for (int i = 0; i < count; i++) {
+        ::dsn::blob key;
+        if (requests[i] != nullptr) {
```

Comment:
When will requests[i] be nullptr?

Reply:
https://github.com/XiaoMi/pegasus/blob/d73a0238681b3a43d7e22d4c4428d87296f2f966/src/server/pegasus_server_write.cpp#L82
When writes failed

Reply:
It's a dassert, means it's invalid and forbiden case.